modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Cursor.scala

```diff
@@ -1,275 +1,243 @@
+    override def focus: Option[Node] = None
 
-    override def downSet: Set[Cursor] =
-      obj.map(new NodeCursorArrayElem(subject, _, this)(this, DownSet, graph))
-
-    override def downAt(o: Node => Boolean): Cursor =
-      obj.find(o) match {
-        case None       => fail(DownAt(o))
-        case Some(node) => new NodeCursorArrayElem(subject, node, this)(this, DownAt(o), graph)
+    override def values: Option[Set[Node]] = {
+      val nodes = cursorSet.foldLeft(Set.empty[Node]) {
+        case (set, c) =>
+          c.focus match {
+            case Some(n) => set + n
+            case None    => set
+          }
       }
+      Some(nodes)
+    }
 
-    def focus: Option[Node]                           = None
-    def values: Option[Iterable[Node]]                = Some(obj)
-    def addOp(cursor: SCursor, op: CursorOp): SCursor = new SetNodeCursorSel(subject, obj, parent)(cursor, op, graph)
-    def downField(p: IriNode => Boolean): Cursor      = fail(DownField(p))
-    def up: Cursor                                    = parent.addOp(this, MoveUp)
-
-  }
-
-  private[rdf] final class ListNodeCursorSel(subject: Node, obj: List[Node], parent: SCursor)(
-      lastCursor: SCursor,
-      lastOp: CursorOp,
-      graph: Graph
-  ) extends FieldCursor(lastCursor, lastOp, graph) {
-
-    def field(p: IriNode => Boolean): Cursor = field(subject, p, parent)
-
-    override def downList: List[Cursor] =
-      obj.map(new NodeCursorArrayElem(subject, _, this)(this, DownList, graph))
-
-    override def downSet: Set[Cursor] =
-      downList.toSet
+    override def cursors: Option[Set[Cursor]] =
+      Some(cursorSet.asInstanceOf[Set[Cursor]])
 
-    override def downAt(o: Node => Boolean): Cursor =
-      obj.find(o) match {
-        case None       => fail(DownAt(o))
-        case Some(node) => new NodeCursorArrayElem(subject, node, this)(this, DownAt(o), graph)
+    override def narrow: Cursor =
+      cursorSet.take(2).toList match {
+        case c :: Nil => c
+        case _        => FailedCursor(this, CursorOp.Narrow, g)
       }
 
-    def focus: Option[Node]                           = None
-    def values: Option[Iterable[Node]]                = Some(obj)
-    def addOp(cursor: SCursor, op: CursorOp): SCursor = new ListNodeCursorSel(subject, obj, parent)(cursor, op, graph)
-    def downField(p: IriNode => Boolean): Cursor      = fail(DownField(p))
-    def up: Cursor                                    = parent.addOp(this, MoveUp)
-
-  }
+    override def succeeded: Boolean = true
 
-  private[rdf] final class NodeCursorArrayElem(subject: Node, obj: Node, parent: SCursor)(
-      lastCursor: SCursor,
-      lastOp: CursorOp,
-      graph: Graph
-  ) extends FieldCursor(lastCursor, lastOp, graph) {
+    override def up(p: IriNode): Cursor = {
+      val cs = cursorSet.foldLeft(Set.empty[NodeCursor]) {
+        case (set, c) =>
+          val ss = g.selectReverse(c.node, p)
+          ss.take(2).toList match {
```

Comment:
This trick is used again and again. I think it can be refactored out on a method

Reply:
I thought about that, but it reduces readability... I don't think it's worth it.